fs/fuse/file.c

```diff
@@ -1395,6 +1397,149 @@ static void fuse_dio_unlock(struct kiocb *iocb, bool exclusive)
 	}
 }
 
+static bool fuse_dlm_locked(struct file* file, loff_t offset, size_t length)
+{
+	struct inode *inode = file_inode(file);
+	struct fuse_conn *fc = get_fuse_conn(inode);
+	struct fuse_inode *fi = get_fuse_inode(inode);
+	struct dlm_locked_area *area;
+
+	/* if dlm is not supported by fuse server, don't bother */
+	if (fc->no_dlm)
+		return true;
+
+	/* check the locked areas for the given offset and length */
+	list_for_each_entry(area, &fi->dlm_locked_areas, list) {
+		loff_t current_area_start_offset = area->offset;
+		loff_t current_area_end_offset = area->offset + area->size;
+		loff_t lock_end_offset = offset + length;
+		loff_t lock_start_offset = offset;
+
+		/* check if the locked areas are completely distinct, then we should continue */
+		if (current_area_end_offset < lock_start_offset
+			|| current_area_start_offset > lock_end_offset)
+			continue;
```

Comment:
The problem with lists is that that random writes to large files will be really slow, I think